integration/log_integration_test.go

```diff
@@ -0,0 +1,332 @@
+		t.Fatalf("Log consistency check failed: %v", err)
+	}
+}
+
+func queueLeaves(treeID int64, client trillian.TrillianLogClient, params testParameters) error {
+	leaves := []trillian.LogLeaf{}
+
+	for l := int64(0); l < params.leafCount; l++ {
+		// Leaf data based on the sequence number so we can check the hashes
+		leafNumber := params.startLeaf + l
+
+		data := []byte(fmt.Sprintf("Leaf %d", leafNumber))
+		hash := sha256.Sum256(data)
+
+		leaf := trillian.LogLeaf{
+			// TODO(Martin2112): This should be LeafValueHash but it doesn't exist yet
+			MerkleLeafHash: hash[:],
+			LeafValue:      data,
+			ExtraData:      nil,
+			LeafIndex:      0}
+		leaves = append(leaves, leaf)
+
+		if len(leaves) >= params.queueBatchSize || (l + 1) == params.leafCount {
+			glog.Infof("Queueing %d leaves ...", len(leaves))
+
+			req := makeQueueLeavesRequest(treeID, leaves)
+			ctx, cancelFunc := getRPCDeadlineContext()
+			response, err := client.QueueLeaves(ctx, &req)
+			cancelFunc()
+
+			if err != nil {
+				return err
+			}
+
+			if response.Status == nil || response.Status.StatusCode != trillian.TrillianApiStatusCode_OK {
+				return fmt.Errorf("queue leaves failed: %s %d", response.Status.Description, response.Status.StatusCode)
+			}
+
+			leaves = leaves[:0] // starting new batch
+		}
+	}
+
+	return nil
+}
+
+func waitForSequencing(treeID int64, client trillian.TrillianLogClient, params testParameters) error {
+	endTime := time.Now().Add(params.sequencingWaitTotal)
+
+	glog.Infof("Waiting for sequencing until: %v", endTime)
+
+	for endTime.After(time.Now()) {
+		req := trillian.GetSequencedLeafCountRequest{LogId: treeID}
+		ctx, cancelFunc := getRPCDeadlineContext()
+		sequencedLeaves, err := client.GetSequencedLeafCount(ctx, &req)
+		cancelFunc()
+
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("Leaf count: %d", sequencedLeaves.LeafCount)
+
+		if sequencedLeaves.LeafCount >= params.leafCount + params.startLeaf {
+			return nil
+		}
+
+		glog.Infof("Leaves sequenced: %d. Still waiting ...", sequencedLeaves.LeafCount)
+
+		time.Sleep(params.sequencingPollWait)
+	}
+
+	return errors.New("wait time expired")
+}
+
+func readbackLogEntries(logID int64, client trillian.TrillianLogClient, params testParameters) (map[int64]*trillian.LogLeaf, error) {
+	currentLeaf := int64(0)
+	leafMap := make(map[int64]*trillian.LogLeaf)
+
+	// Build a map of all the leaf data we expect to have seen when we've read all the leaves.
+	// Have to work with strings because slices can't be map keys. Sigh.
+	leafDataPresenceMap := make(map[string]bool)
+
+	for l := int64(0); l < params.leafCount; l++ {
+		leafDataPresenceMap[fmt.Sprintf("Leaf %d", l + params.startLeaf)] = true
+	}
+
+	for currentLeaf < params.leafCount {
+		hasher := merkle.NewRFC6962TreeHasher(crypto.NewSHA256())
+
+		// We have to allow for the last batch potentially being a short one
+		numLeaves := params.leafCount - currentLeaf
+
+		if numLeaves > int64(params.readBatchSize) {
```

Comment:
You could add a `bs := int64(params.readBatchSize)` up the top to save having to cast twice here, or, alternatively, make readBatchSizeFlag an int64?